slackbot.js

```diff
@@ -9,49 +9,52 @@ function withDefaultOptions (env, options) {
 
 function Slackbot (transport, env, options) {
   var mergedOptions = withDefaultOptions(env, options);
-  var rtm = new transport.RtmClient(env.vars.SLACK_TOKEN, mergedOptions);
-  var web = new transport.WebClient(env.vars.SLACK_TOKEN, mergedOptions);
-  var logger = helpers.logger('Slackbot', mergedOptions.logLevel === 'verbose');
+  this.rtm = new transport.RtmClient(env.vars.SLACK_TOKEN, mergedOptions);
+  this.web = new transport.WebClient(env.vars.SLACK_TOKEN, mergedOptions);
+  this.logger = helpers.logger('Slackbot', mergedOptions.logLevel === 'verbose');
+}
 
-  function includeChannel (next) {
-    var yak = logger.sub('includeChannel');
-    return function addChannelToResponse (response) {
+Slackbot.prototype.includeChannel = function (next) {
+  var yak = this.logger.sub('includeChannel');
+  return (function (web) {
+    return function (response) {
       if (!response.channel_id) {
         yak('no channel_id found, will not place call', response);
         return next(response);
       }
       yak('channel_id found in response, get info for', response.channel_id);
       return web.channels.info(
         response.channel_id,
-        helpers.handleError(function (channelRes) {
-          yak('channel response for', response.channel_id, channelRes);
-          next(response, channelRes);
+        helpers.handleError(function (channelResponse) {
+          yak('channel response for', response.channel_id, channelResponse);
+          next(response, channelResponse);
         }));
     };
-  }
-  function getFileInfo (id, next) {
-    var yak = logger.sub('getFileInfo');
-    yak('getting file', id);
-    web.files.info(id, helpers.handleError(function (res) {
-      yak('received file', res.file);
-      next(res.file);
-    }));
-  }
-  function onPinAdded (next) {
-    rtm.on(slack.RTM_EVENTS.PIN_ADDED, includeChannel(function (res, chan) {
-      logger('onPinAdded', res);
-      next.call(this, res, chan);
-    }));
-  }
-  function connect (next) {
-    rtm.on(slack.CLIENT_EVENTS.RTM.AUTHENTICATED, next);
-    rtm.start();
-  }
-  return {
-    getFileInfo: getFileInfo,
-    onPinAdded: onPinAdded,
-    connect: connect
-  };
-}
+  })(this.web);
+};
+
+Slackbot.prototype.getFileInfo = function (id, next) {
+  var yak = this.logger.sub('getFileInfo');
+  yak('getting file', id);
+  this.web.files.info(id, helpers.handleError(function (res) {
+    yak('received file', res.file);
+    next(res.file);
+  }));
+};
+
+Slackbot.prototype.onPinAdded = function (next) {
+  // TODO verify use of 'this' in next.call
+  // TODO verify change in logger logic
```

Comment:
the reason you were getting an error trying to call `this.logger('onPinAdded')` here is that the value of `this` changed once you were inside this anonymous function. (this is one of the pitfalls of using prototypes and `this` over closure). looks like you figured that out and did the most common fix one does for that by creating and closing over a reference. however, in this implementation, `next` will still be called with the wrong `this`. i'm not sure if that's actually a bug, since i don't see a reason for the callback to run with `this` as the bot instance. anyway, there are a couple of ways to go with this general problem.

#### Function.prototype.bind
```javascript
this.rtm.on(slack.RTM_EVENTS.PIN_ADDED, this.includeChannel(function (response, channelResponse) {
  this.logger('onPinAdded', response);
  next.call(this, response, channelResponse);
}.bind(this)));
```

#### ES6 arrow functions
```javascript
this.rtm.on(slack.RTM_EVENTS.PIN_ADDED, this.includeChannel((response, channelResponse) => {
  this.logger('onPinAdded', response);
  next.call(this, response, channelResponse);
}));
```

i recommend the latter, i use ES6 arrows everywhere (except prototype methods, where you WANT `this` to be late bound instead of lexically bound). i didn't in this project just for clarity's sake.